extensions/opencv/MotionDetector.cpp

```diff
@@ -0,0 +1,222 @@
+  cv::findContours(thresh, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
+
+  // Finish process
+  thresh.copyTo(diff);
+
+  bool moved = false;
+  for(const auto &contour : contours) {
+    auto area = cv::contourArea(contour);
+    if (area < min_area_) {
+      continue;
+    }
+    // TODO - Nghia: If a contour exists for >n frames, then we should consider it as background
+    // TODO - Nghia: How? Map<rectangle, cnt>?
+    moved = true;
+    cv::Rect bbox = cv::boundingRect(contour);
+    cv::rectangle(frame, bbox.tl(), bbox.br(), cv::Scalar(0, 255, 0), 1, 8, 0);
+  }
+  if (!moved) {
+    // Adaptive background, update background so that the illumnation does not affect that much.
+    cv::accumulateWeighted(gray, background_, 0.5);
+    cv::convertScaleAbs(background_, bg_img_);
+  } else if (update_bg_) {
+    // Use the previous frame as background.
+    gray.copyTo(bg_img_);
+  }
+  return moved;
+}
+
+void MotionDetector::onTrigger(const std::shared_ptr<core::ProcessContext> &context,
+                                 const std::shared_ptr<core::ProcessSession> &session) {
+  std::unique_lock<std::mutex> lock(mutex_, std::try_to_lock);
+  if (!lock.owns_lock()) {
+    logger_->log_info("Cannot process due to an unfinished onTrigger");
+    context->yield();
+    return;
+  }
+
+  auto flow_file = session->get();
+  if (!flow_file) {
+    logger_->log_info("Empty flow file");
+    return;
+  }
+  cv::Mat frame;
+
+  FrameReadCallback cb(&frame);
+  session->read(flow_file, &cb);
+
+  if (frame.empty()) {
+    logger_->log_error("Empty frame.");
+    session->transfer(flow_file, Failure);
+  }
+
+  // hardcoded width to 500
+  double scale = float(500)/frame.size().width;
+  cv::resize(frame, frame, cv::Size(0, 0), scale, scale);
+
+  if (background_.empty()) {
+    logger_->log_info("Background is missing, update and exit.");
+    cv::cvtColor(frame, bg_img_, cv::COLOR_BGR2GRAY);
+    cv::GaussianBlur(bg_img_, bg_img_, cv::Size(21, 21), 0, 0);
+//    cv::equalizeHist(bg_img_, bg_img_);
+    bg_img_.convertTo(background_, CV_32F);
+    return;
+  }
+  logger_->log_info("Start motion detecting");
+
+  cv::Mat img_diff;
+  auto diff_file = session->create();
+
+  auto t = std::time(nullptr);
+  auto tm = *std::localtime(&t);
+  std::ostringstream oss;
+  oss << std::put_time(&tm, "%d-%m-%Y %H-%M-%S");
+  auto filename = oss.str();
+  auto filename2 = filename + "_diff";
+  filename.append(image_encoding_);
+  filename2.append(image_encoding_);
+
+  detectAndDraw(frame, img_diff);
+
+  FrameWriteCallback write_cb(frame, image_encoding_);
+  FrameWriteCallback write_cb2(img_diff, image_encoding_);
+
+  session->putAttribute(flow_file, "filename", filename);
+  session->putAttribute(diff_file, "filename", filename2);
+
+  session->write(flow_file, &write_cb);
+  session->write(diff_file, &write_cb2);
+  session->transfer(flow_file, Success);
+  session->transfer(diff_file, Success);
```

Comment:
I'm not sure if these two should be transferred to the same relationship. Do we need the original FF anyway?